table/block_based_table_reader.cc

```diff
@@ -1230,32 +1235,35 @@ Status BlockBasedTable::GetDataBlockFromCache(
 
   // found compressed block
   RecordTick(statistics, BLOCK_CACHE_COMPRESSED_HIT);
-  compressed_block = reinterpret_cast<Block*>(
+  compressed_block = reinterpret_cast<BlockContents*>(
       block_cache_compressed->Value(block_cache_compressed_handle));
-  assert(compressed_block->compression_type() != kNoCompression);
+  CompressionType compression_type = static_cast<CompressionType>(
+      compressed_block->data.data()[compressed_block->data.size()]);
+  assert(compression_type != kNoCompression);
 
   // Retrieve the uncompressed contents into a new buffer
   BlockContents contents;
-  UncompressionContext uncompresssion_ctx(compressed_block->compression_type(),
-                                          compression_dict);
-  s = UncompressBlockContents(uncompresssion_ctx, compressed_block->data(),
-                              compressed_block->size(), &contents,
-                              format_version, ioptions, allocator);
+  UncompressionContext uncompresssion_ctx(compression_type, compression_dict);
+  s = UncompressBlockContents(uncompresssion_ctx, compressed_block->data.data(),
+                              compressed_block->data.size(), &contents,
+                              rep->table_options.format_version, rep->ioptions,
+                              GetMemoryAllocator(rep->table_options));
 
   // Insert uncompressed block into block cache
   if (s.ok()) {
     block->value =
-        new Block(std::move(contents), compressed_block->global_seqno(),
+        new Block(std::move(contents), rep->get_global_seqno(is_index),
                   read_amp_bytes_per_bit,
                   statistics);  // uncompressed block
-    assert(block->value->compression_type() == kNoCompression);
-    if (block_cache != nullptr && block->value->cachable() &&
+    if (block_cache != nullptr && block->value->own_bytes() &&
```

Comment:
There are 3 other block_cache->Insert statements in this file (for index and filters). How come we do not check own_bytes() before inserting them into block cache?

Reply:
It's not new code. It's just a mapping from previous cachable() call. The three other places are for index, filter block and dummy costing.

Reply:
I understand it is not new code but as we are doing a refactoring now, it is a good time to understand the old code better. If the the source does not own bytes, like in mmap files, why should we add them to the block cache? Lets say we add an index block from mmap file to block cache and them later the corresponding memory of mmap file is overwritten but we keep reading from it since the block cache refers to it.